Sources/Signal.swift

```diff
@@ -2260,23 +2260,55 @@ extension Signal {
 	///   - error: Error to send with failed event if `self` is not completed
 	///            when `interval` passes.
 	///   - interval: Number of seconds to wait for `self` to complete.
-	///   - scheudler: A scheduler to deliver error on.
+	///   - scheduler: A scheduler to deliver error on.
 	///
 	/// - returns: A signal that sends events for at most `interval` seconds,
 	///            then, if not `completed` - sends `error` with failed event
 	///            on `scheduler`.
 	public func timeout(after interval: TimeInterval, raising error: Error, on scheduler: DateScheduler) -> Signal<Value, Error> {
+		return timeout(after: interval, sending: .failed(error), on: scheduler)
+	}
+
+	/// Forward events from `self` until `interval`. Then if signal isn't
+	/// completed yet, sends `event` and terminates.
+	///
+	/// - note: If `event` is a value, the returned signal will automatically
+	///         complete after sending it.
+	///
+	/// - note: If the interval is 0, the timeout will be scheduled immediately.
+	///         The signal must complete synchronously (or on a faster
+	///         scheduler) to avoid the timeout.
+	///
+	/// - precondition: `interval` must be non-negative number.
+	///
+	/// - parameters:
+	///   - event: An event to terminate with if `self` is not completed
+	///            when `interval` passes.
+	///   - interval: Number of seconds to wait for `self` to complete.
+	///   - scheduler: A scheduler to deliver error on.
+	///
+	/// - returns: A signal that sends events for at most `interval` seconds,
+	///            then, if not `completed` - sends `error` with failed event
+	///            on `scheduler`.
+	public func timeout(after interval: TimeInterval, sending event: Event, on scheduler: DateScheduler) -> Signal<Value, Error> {
 		precondition(interval >= 0)
 
 		return Signal { observer in
 			let disposable = CompositeDisposable()
 			let date = scheduler.currentDate.addingTimeInterval(interval)
+			var signalDisposable: Disposable?
 
 			disposable += scheduler.schedule(after: date) {
-				observer.send(error: error)
+				signalDisposable?.dispose()
```

Comment:
Hmm, it is a legit race if we guarantee that the `value` sent by `timeout` must be followed by a terminal event. You might want to use `SerialDisposable` for `signalDisposable` then, since statically it is possible that the scheduled block runs before the store of the observer handle.